R/stats_utils.R

```diff
@@ -0,0 +1,296 @@
+  days.threshold <- 359
+  check.basic.argument.validity(tmax, tmin, prec, tmax.dates, tmin.dates, prec.dates, base.range, n)
+  
+  d.list <- list(tmin.dates, tmax.dates, prec.dates)
+  all.dates <- do.call(c, d.list[!sapply(d.list, is.null)])
+  last.day.of.year <- get.last.monthday.of.year(all.dates)
+  cal <- attr(all.dates, "cal")
+  
+  bs.date.range <- as.PCICt(paste(base.range, c("01-01", last.day.of.year), sep="-"), cal=cal)
+  new.date.range <- as.PCICt(paste(as.numeric(format(range(all.dates), "%Y", tz="GMT")), c("01-01", last.day.of.year), sep="-"), cal=cal)
+  date.series <- seq(new.date.range[1], new.date.range[2], by="day")
+  bs.date.series <- seq(bs.date.range[1], bs.date.range[2], by="day")
+  
+  quantiles <- list()
+  
+  if(!is.null(tmax)) {
+    if(get.num.days.in.range(tmax.dates, bs.date.range) <= days.threshold)
+      stop("There is less than a year of tmax data within the base period. Consider revising your base range and/or check your input data.")
+    filled.tmax <- create.filled.series(tmax, trunc(tmax.dates, "days"), date.series)
+    quantiles$tmax <- get.temp.var.quantiles(filled.tmax, date.series, bs.date.series, temp.qtiles, bs.date.range, n)
+  } 
+  
+  if(!is.null(tmin)) {
+    if(get.num.days.in.range(tmin.dates, bs.date.range) <= days.threshold)
+      stop("There is less than a year of tmin data within the base period. Consider revising your base range and/or check your input data.")
+    filled.tmin <- create.filled.series(tmin, trunc(tmin.dates, "days"), date.series)
+    quantiles$tmin <- get.temp.var.quantiles(filled.tmin, date.series, bs.date.series, temp.qtiles, bs.date.range, n)
+  }
+  
+  if(!is.null(prec)) {
+    if(get.num.days.in.range(prec.dates, bs.date.range) <= days.threshold)
+      stop("There is less than a year of prec data within the base period. Consider revising your base range and/or check your input data.")
+    filled.prec <- create.filled.series(prec, trunc(prec.dates, "days"), date.series)
+    quantiles$prec <- get.prec.var.quantiles(filled.prec, date.series, bs.date.range, prec.qtiles)
+  }
+  return(quantiles)
+}
+
+# Computes a specified statistic (min, max) for a given climate index and frequency.
+compute.stat <- function(ci, stat, data.key, freq = c("monthly", "annual", "seasonal"), include.exact.dates) {
+  stopifnot(!is.null(ci@data[[data.key]]))
+  data <- ci@data[[data.key]]
+  date.factors <- ci@date.factors[[match.arg(freq)]]
+  mask <- ci@namasks[[match.arg(freq)]][[data.key]]
+  cal <- attr(ci@dates, "cal")
+  
+  if (include.exact.dates) {
+    return(exact.date(stat, data, date.factors, freq, cal, mask))
+  }
+  
+  return(suppressWarnings(tapply.fast(data, date.factors, stat, na.rm = TRUE)) * mask)
+}
+
+#' Get available indices by name
```

Comment:
This function seems like kind of the odd one out in this file, since it's not just about statistics but about _all_ the indices. Do you think it would make sense in climdexInput_class.R? It evaluates an input object, right? Similar to the checks? That file is already the biggest one, though....

Reply:
Good point, my initial vision for climdexInput was to have a separate file for checks and validation. Breaking it up that way would give the `get.available.indices` function a better home AND reduce the  `climdexInput_class` file size.